source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,163 @@
+  }
+  for (const auto& param : it->second.from_params()) {
+    extract_param_.param_keys_.insert(param);
+  }
+  if (it->second.from_headers_size() == 0 && it->second.from_params_size() == 0) {
+    extract_param_.header_keys_.insert("authorizationBearer ");
+  }
+  issuer_ = it->second.issuer();
+}
+
+void ProviderNameVerifier::verify(Http::HeaderMap& headers, Verifier::Callbacks& callback) {
+  beforeVerify(audiences_, headers, callback);
+
+  auth_->verify(&extract_param_, absl::optional<std::string>{issuer_}, headers, this);
+}
+
+void ProviderNameVerifier::onComplete(const Status& status) {
+  if (callback_) {
+    callback_->onComplete(status);
+    callback_ = nullptr;
+    close();
+  }
+}
+
+void AllowFailedVerifier::verify(Http::HeaderMap& headers, Verifier::Callbacks& callback) {
+  beforeVerify({}, headers, callback);
+
+  auth_->verify(nullptr, absl::nullopt, headers, this);
+}
+
+void AllowFailedVerifier::onComplete(const Status&) {
+  if (callback_) {
+    callback_->onComplete(Status::Ok);
+    callback_ = nullptr;
+    close();
+  }
+}
+
+void GroupVerifier::verify(Http::HeaderMap& headers, Verifier::Callbacks& callback) {
+  callback_ = &callback;
+  count_ = 0;
+  for (const auto& it : verifiers_) {
+    if (callback_) {
+      it->verify(headers, *this);
+    }
+  }
+}
+
+void GroupVerifier::close() {
+  for (const auto& it : verifiers_) {
+    it->close();
+  }
+}
+
+AnyVerifier::AnyVerifier(const JwtRequirementOrList& or_list, const AuthFactory& factory,
+                         const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers) {
+  for (const auto& it : or_list.requirements()) {
+    verifiers_.push_back(Verifier::create(it, providers, factory));
+  }
+}
+
+void AnyVerifier::onComplete(const ::google::jwt_verify::Status& status) {
+  if (!callback_) {
+    return;
+  }
+  if (++count_ == verifiers_.size() || Status::Ok == status) {
+    callback_->onComplete(status);
+    callback_ = nullptr;
+  }
+}
+
+AllVerifier::AllVerifier(const JwtRequirementAndList& and_list, const AuthFactory& factory,
+                         const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers) {
+  for (const auto& it : and_list.requirements()) {
+    verifiers_.push_back(Verifier::create(it, providers, factory));
+  }
+}
+
+void AllVerifier::onComplete(const ::google::jwt_verify::Status& status) {
+  if (!callback_) {
+    return;
+  }
+  if (++count_ == verifiers_.size() || Status::Ok != status) {
+    callback_->onComplete(status);
+    callback_ = nullptr;
+  }
+}
+
+VerifierPtr Verifier::create(const JwtRequirement& requirement,
+                             const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                             const AuthFactory& factory) {
+  switch (requirement.requires_type_case()) {
+  case JwtRequirement::RequiresTypeCase::kProviderName:
+    return std::make_unique<ProviderNameVerifier>(requirement.provider_name(),
```

Comment:
Since we already know the provider_name,  should we just lookup the provider_map, pass the issuer info into the object to use.

Reply:
Yes. I can pass in the JwtProvider, as I need token header and param locations along with issuer.